internal/orchestrator/orchestrator.go

```diff
@@ -89,56 +107,115 @@ func (o *Orchestrator) Run(ctx context.Context) {
 		case <-ctx.Done():
 			o.streamBroker.Close()
 			return
+
+		// retrieve and process events sent by controllers.
+		case <-tickerFetchEvents:
+			if o.concurrencyLimit() {
+				continue
+			}
+
+			o.pullEvents(ctx)
+
+		// eventsCh receives events from push based subscriptions, in this case serverservice.
 		case msg := <-eventsCh:
-			o.processMsg(ctx, msg)
+			o.processEvent(ctx, msg)
 		}
 	}
 }
 
-func (o *Orchestrator) processMsg(ctx context.Context, msg events.Message) {
-	otelCtx, span := otel.Tracer(pkgName).Start(ctx, "processMsg")
-	defer span.End()
+func (o *Orchestrator) concurrencyLimit() bool {
+	return int(o.dispatched) >= o.concurrency
+}
 
-	data, err := msg.Data()
+func (o *Orchestrator) pullEvents(ctx context.Context) {
+	// XXX: consider having a separate context for message retrieval
+	msgs, err := o.streamBroker.PullMsg(ctx, 1)
 	if err != nil {
 		o.logger.WithFields(
-			logrus.Fields{"err": err.Error(), "subject": msg.Subject()},
-		).Error("data unpack error")
+			logrus.Fields{"err": err.Error()},
+		).Debug("error fetching work")
+	}
+
+	for _, msg := range msgs {
+		if ctx.Err() != nil || o.concurrencyLimit() {
+			o.eventNak(msg)
+
+			return
+		}
+
+		// spawn msg process handler
+		o.syncWG.Add(1)
```

Comment:
Nit/Food For Thought: shouldn't we increment by `len(msgs)`? Otherwise when the call to `PullMsg` is updated to `n>1` we'd have `concurrencyLimit` goroutines handling `len(msgs)` messages?

Reply:
I was trying to have the waitgroup `Add(1)` as close to the spawned routine just so returned errors don't cause an incorrect waitgroup increment. 
Can you elaborate on the second part of your question - `when the call to PullMsg is updated to n>1 we'd have concurrencyLimit goroutines handling len(msgs) messages?`

Reply:
As currently coded, the `PullMsg` method can pull a batch of messages created then 1. My question is: is this concurrency limit for limiting go-routines or messages-in-flight? What is the rationale for the limit?

Reply:
Yes, the limit was introduced mainly so we can manage how many concurrent Goroutines are spawned, we can experiment with increasing it linearly based on how we see the Orchestrator handling messages.